container-core/src/main/java/com/yahoo/restapi/RestApi.java

```diff
@@ -0,0 +1,367 @@
+        final Class<EXCEPTION> type;
+        final ExceptionMapper<EXCEPTION> mapper;
+
+        ExceptionMapperHolder(Class<EXCEPTION> type, ExceptionMapper<EXCEPTION> mapper) {
+            this.type = type;
+            this.mapper = mapper;
+        }
+
+        boolean matches(RuntimeException e) { return type.isAssignableFrom(e.getClass()); }
+        HttpResponse toResponse(RuntimeException e, RequestContext context) { return mapper.toResponse(type.cast(e), context); }
+    }
+
+    private static class ResponseMapperHolder<ENTITY> {
+        final Class<ENTITY> type;
+        final ResponseMapper<ENTITY> mapper;
+
+        ResponseMapperHolder(Class<ENTITY> type, ResponseMapper<ENTITY> mapper) {
+            this.type = type;
+            this.mapper = mapper;
+        }
+
+        boolean matches(Object entity) { return type.isAssignableFrom(entity.getClass()); }
+        HttpResponse toHttpResponse(Object entity, RequestContext context) { return mapper.toHttpResponse(type.cast(entity), context); }
+    }
+
+    public static class Builder {
+        private final List<Route> routes = new ArrayList<>();
+        private final List<ExceptionMapperHolder<?>> exceptionMappers = new ArrayList<>();
+        private final List<ResponseMapperHolder<?>> responseMappers = new ArrayList<>();
+        private Route defaultRoute;
+        private ObjectMapper jacksonJsonMapper;
+
+        private Builder() {}
+
+        public Builder setObjectMapper(ObjectMapper mapper) { this.jacksonJsonMapper = mapper; return this; }
+
+        public Builder setDefaultRoute(Route route) { this.defaultRoute = route; return this; }
+
+        public Builder addRoute(Route route) { routes.add(route); return this; }
+
+        public <EXCEPTION extends RuntimeException> Builder addExceptionMapper(Class<EXCEPTION> type, ExceptionMapper<EXCEPTION> mapper) {
+            exceptionMappers.add(new ExceptionMapperHolder<>(type, mapper)); return this;
+        }
+
+        public <ENTITY> Builder addResponseMapper(Class<ENTITY> type, ResponseMapper<ENTITY> mapper) {
+            responseMappers.add(new ResponseMapperHolder<>(type, mapper)); return this;
+        }
+
+        public RestApi build() { return new RestApi(this); }
+    }
+
+    public static class Route {
+        private final String pathPattern;
+        private final Map<Method, MethodHandler<?>> handlerPerMethod;
+        private final MethodHandler<?> defaultHandler;
+
+        private Route(Builder builder) {
+            this.pathPattern = builder.pathPattern;
+            this.handlerPerMethod = Map.copyOf(builder.handlerPerMethod);
+            this.defaultHandler = builder.defaultHandler != null ? builder.defaultHandler : createDefaultMethodHandler();
+        }
+
+        private static MethodHandler<?> createDefaultMethodHandler() {
+            return context -> { throw new MethodNotAllowedException(context.request()); };
+        }
+
+        public static class Builder {
```

Comment:
This is the second builder in this class, that's a bit confusing.  More generally RestApi contains a lot of inner classes and interfaces, making e.g. navigation in this file a bit confusing.  Consider moving all inner classes and interfaces of RestApi to the top-level, except the builder of RestApi.

Reply:
Good point! All private types could be moved out to top level. All classes/interfaces that are public must either be modelled as inner classes/interfaces or moved to new source files (as a Java source files cannot contain multiple public classes at top level).